### ISSUE TITLE
Umbraco .net core implementation should support embed db, which is multiplatform

### ISSUE THREAD

**Issue by bielu** [umbraco/Umbraco-CMS#7831]
SQL CE is not supported on other platforms than windows and it is actually really bad for migration to .net core. I think Umbraco should consider support for some embed db different than sql ce or LocalDB. Two examples:
https://www.litedb.org/ (I know it is nosql, but still could be considered) 
https://www.sqlite.org/index.html
I know it is a lof effort to move that to any of that DBs but it will be really beneficial for migration to .net core as it will be one fewer windows dependency.

**Comment by nul800sebastiaan** [comment]
Thanks, for now we're going to still support SQL CE and SQL Server - https://github.com/umbraco/rfcs/blob/cba1b60348757a0cb85cd694705d85ae51993ae0/cms/0000-project-unicore-structure.md#umbracopersistencesqlce

It will be possible to build other providers, which is on our wishlist, but that's not going to make it into the initial implementation.

**Comment by nul800sebastiaan** [comment]
On a sidenote: SQL Server is available for Linux as well, which makes it one fewer windows dependency :-)

**Comment by bielu** [comment]
On sidenote: even if SQl Server is available on Linux, then if Umbraco is going to attach SQL ce dll on .net it means we actually blocking the possibility of using Umbraco on Linux or any another platform... Maybe it is just me but I don't understand to stay with sql CE, which has EOL  July 2021. It is actually same not a good type of decision as staying on angular js for backoffice.

**Comment by nul800sebastiaan** [comment]
In any case, running on anything but Windows is not prioritized right now. The SQL CE implementation is a separate dll so can be decoupled if necessary.

We are well aware of EOL dates of several components we're using. :-) 
SQL CE will still be around for a while because it is something we're very familiar with, we have the code that makes it compatible and it is not broken or compromised. 

We'll move away to something else in due time, when we have time to work on alternatives. This might even be before a final release of vNext, but we can't commit to that at the moment.

**Comment by bielu** [comment]
> In any case, running on anything but Windows is not prioritized right now.

and this is a thing which I don't understand really, it should be a priority as more and more people are interested in .net core as it is cross-platform and we are actually shooting in own knee without support to other platforms

> The SQL CE implementation is a separate dll so can be decoupled if necessary.

Umbraco shouldn't really care about that... 

> SQL CE will still be around for a while because it is something we're very familiar with, we have the code that makes it compatible and it is not broken or compromised.

this is the perfect time to remove that as .net core implementation should introduce a lot breaking changes to remove most of the legacy code...


> This might even be before a final release of vNext, but we can't commit to that at the moment.

If it is not we are repeating scenario from v5...

**Comment by nul800sebastiaan** [comment]
Your ideas are perfectly valid, but we currently don't have the resources to put them in. We'll get there in due time. Eventually we'll be cross platform, eventually we'll use different types of database, eventually we'll replace AngularJS, etc. etc.

> If it is not we are repeating scenario from v5...

Completely the opposite.. we're not over-engineering, we're not drastically changing everything, we're not building a huge system that doesn't work until a final big bang release.
By working in small increments as we are doing now, we're avoiding those mistakes we made for v5.

**Comment by bielu** [comment]
> Completely the opposite.. we're not over-engineering, we're not drastically changing everything, we're not building a huge system that doesn't work until a final big bang release.

You are building solution which will have big expectation but will not fit them at all, as people are hoping to work with Umbraco on Linux/MacOs/etc...

**Comment by nul800sebastiaan** [comment]
> Umbraco on Linux/MacOs/etc...

We'll get there. 🙂

**Comment by bielu** [comment]
> We'll get there. 🙂

Question is when

**Comment by aaron-sadler-activeis** [comment]
I want to add, the lack of support for Linux is pretty disappointing, as this one of the considerations when considering dot net core.

I would expect Umbraco to run on Linux, in the case using of a MSSQL database.

And NuGet packages which are .netstandard should also run on Linux, so other than an extremly outdated and unsupported database engine (SQL CE), I can't see the issue...

**Comment by Shazwazza** [comment]
Let's not dwell on Sqlce and non-linux support isn't actually a thing. So let me rephrase:

Our focus is on MSSQL and we will provide LocalDb support too. If possible, we will also still support SqlCe on Windows... but we aren't at a place where we can see if that is even possible yet. It is not our priority to make it work if it's doesn't 'just' work as it is today.

Our priority during development on netcore is not on Linux support. That doesn't mean it won't work on Linux... because netcore works there. But does it work today? No because we aren't even close to that stage yet.

At some stage we'll probably have an SQLite embedded DB, but again, this is not a priority right now and won't be until we have things working on MSSQL on windows and then we will explore getting other things working. 

Making more sense?

**Comment by bielu** [comment]
>Our focus is on MSSQL and we will provide LocalDb support too. If possible, we will also still support SqlCe on Windows... but we aren't at a place where we can see if that is even possible yet. It is not our priority to make it work if it's doesn't 'just' work as it is today.

LocalDb works only on windows also..., SQL CE should be dead and not used at all etc.

>Our priority during development on netcore is not on Linux support. That doesn't mean it won't work on Linux... because netcore works there. But does it work today? No because we aren't even close to that stage yet.

But Umbraco has few Windows dependencies, and moving to netcore doesn't mean it will be removed and from that topic, we could feel like they will be not resolved... 

>At some stage we'll probably have an SQLite embedded DB, but again, this is not a priority right now and won't be until we have things working on MSSQL on windows and then we will explore getting other things working.

@Shazwazza but actually mssql for Linux and Windows are fully compatible from the perspective of Umbraco. 

>Making more sense?

Little more, but still not really convinced with Umbraco is going to handle working cross handle as a proposition of using localdb is tied up us again to Windows, as if you are going deliver dll for that it means it will not run on another platform correctly as long you are not going to handle attaching provider based on platform, but don't think so it is a good idea...

**Comment by Shazwazza** [comment]
LocalDb has no extra dependencies and no windows only dependencies. It is just a connection string and it uses the LocalDb SQL server installed on the machine. In no way is there any windows specific code to make this work.

In fact there is no windows specific code at all in the codebase that will be integral to the operation of the CMS. If SqlCe works, that will be a separate package. Again... lets not get bogged down by anything to do with SqlCe, it is an afterthought, it isn't a priority. 

We are building against netstandard and netcore3.1, there are no #if checks for different builds, there are no OS checks at runtime. So there is no reason it won't work cross platform... It will, but like i said, it's not a priority 'now'.

**Comment by bielu** [comment]
> LocalDb has no extra dependencies and no windows only dependencies. It is just a connection string and it uses the LocalDb SQL server installed on the machine. In no way is there any windows specific code to make this work.

LocalDB doesn't work on Linux at all, I was researching that for a private projects.
as example:
https://stackoverflow.com/questions/45860851/localdb-is-not-supported-on-this-platform
https://github.com/dotnet/AspNetCore.Docs.de-de/issues/35
Etc.

**Comment by Shazwazza** [comment]
Yes we know, nobody will be forced to use local db, we can disable that option on install on Linux. Sql server will just work as is.

**Comment by bielu** [comment]
> Yes we know, nobody will be forced to use local db, we can disable that option on install on Linux. Sql server will just work as is.

Then why you want to deliver something that will not work everywhere?!

**Comment by Shazwazza** [comment]
Because It's just a connection string, it's no extra work and it works right now. We also need a DB for netcore integration tests on azure devops and this is working with LocalDB right now. 

> Question is when

Creating a brand new embedded db provider will just prolong a release. Sometime in the future, a cross platform embedded db can be investigated but it's not a priority right now.
